### ISSUE TITLE
Connection blues

### ISSUE THREAD

**Issue by kaipila** [LN-Zap/node-lnd-grpc#69]
Hello,

Banging my head against the wall for a day, I figured I'd drop a line here.

If my initial grpc.connect() fails it seems impossible for me to recover from this.

Any further attempt to try calling grpc.connect() fails griping about an invalid state transition and fails. Can't "connect" from "connected" state or something to that effect. The on-active handler is never called.

I guess the theory is that reconnect is attempted if there is a call on the channel but since I have not made initial contact I have no services available to call to.

The @grpc-node library yours is based on seems to have a grpc.getConnectivityState() that can be asked to initiate a reconnect without services, but I could not get my hands on it in your library.

I got all my credentials in order and I am actually able to make connections under normal circumstances fine, but. 

If I kill my lnd first, try to establish a connection to it and fail, start the lnd up again and boom, I am stuck in a situation where I can not connect to it having had the initial call fail. Just trying to simulate a situation where lnd is not yet running when I start my software up and it comes online later on.

Latest versions of everything (lnd, node, this library) I installed couple of days back.

I don't know, let see how you feel about this =)

**Comment by mrfelton** [comment]
@kaipila this issue has been somewhat addressed in the `next` branch where we use timeouts to catch connection errors and automatically disconnect from any connected services and reset the state machine in the event of a failure.

You could try installing `lnd-grpc@next`.

There are a couple of differences though in the expected call format. Specifically, we no longer automatically connect you to the `Lightning` interface after successfully creating or unlocking a wallet - you must do this manually by calling `activateLightning()` yourself.

**Comment by jacohend** [comment]
@mrfelton Unfortunately I'm getting this "transition is invalid" error even on the `next` branch. Any advice?

```
let lnd = new LndGrpc({
  host: env.LND_SOCKET,
  cert: `/root/.lnd/tls.cert`,
  macaroon: `/root/.lnd/data/chain/bitcoin/${env.NETWORK}/admin.macaroon`
})

async function connectToLndAsync() {
  try {
    await lnd.disconnect()
  } catch (error) {
    logger.error(`LND disconnect failed: ${error.message}`)
  }
  try {
    await lnd.connect()
    logger.info(`wallet password: ${env.HOT_WALLET_PASS}`)
    try {
      await lnd.services.WalletUnlocker.unlockWallet({
        wallet_password: Buffer.from(env.HOT_WALLET_PASS)
      })
    } catch (error) {
      logger.error(`Can't unlock LND, already unlocked? : ${error.message}`)
    }
    await lnd.activateLightning()
    lnd.once('active', async () => {
      logger.info('GRPC state active')
    })
  } catch (error) {
    throw new Error(`Unable to connect to LND : ${error.message}`)
  }
```

**Comment by mrfelton** [comment]
I think the problem here is that you are trying to call `activateLightning`, when that service is already active.

You should verify the service state before trying to initiate state transitions. Try reworking your code something more like this:

```js
  const grpc = new LndGrpc(grpcOptions)
  grpc.once('active', async () => {
    console.info('GRPC state active')
  })

  try {
    await grpc.disconnect()
  } catch (error) {
    console.error(`LND disconnect failed: ${error.message}`)
  }

  try {
    await grpc.connect()
    if (grpc.state === 'locked') {
      try {
        await grpc.services.WalletUnlocker.unlockWallet({
          wallet_password: Buffer.from('password'),
        })
        await grpc.activateLightning()
      } catch (error) {
        console.error(`Can't unlock LND: ${error.message}`)
      }
    }
  } catch (error) {
    throw new Error(`Unable to connect to LND : ${error.message}`)
  }
```

**Comment by mrfelton** [comment]
Note that the initial call to `connect()` will set the correct state based on on the state of the node that you are connecting to, leaving you in in either a `locked` or `active` state.

To illustrate, this is the code of the connect:

```js
  async connect() {
    debug(`Connecting to lnd gRPC service`)

    // Verify that the host is valid.
    const { host } = this.options
    await validateHost(host)

    // Probe the services to determine the wallet state.
    const walletState = await this.determineWalletState()

    // Update our state accordingly.
    switch (walletState) {
      case WALLET_STATE_LOCKED:
        await this.activateWalletUnlocker()
        break

      case WALLET_STATE_ACTIVE:
        await this.activateLightning()
        break
    }
  }
```

So if you connect to an already unlocked node, trying to call `activateLightning` after connecting is going to be invalid.

**Comment by mrfelton** [comment]
Note. The `next` branch has now been merged into the `master` branch and published as `v0.3.0`

**Comment by mrfelton** [comment]
Closing due to inactivity
